cmail-admin/doc/cmail-admin-user.8
.TH CMAIL-ADMIN-USER 8 "August 2015" "v0.1" "cmail Administration Tools"
.SH NAME
cmail-admin-user \- Manage cmail users
.SH SYNOPSIS
.BI "cmail-admin-user [ --dbpath " "master-database-file" " ] [ --verbosity " N
.BI " | -v " N " ] [ --help | -h ] " "mode" " "  "mode-options"
.SH DESCRIPTION
.B cmail-admin-user
allows the administrator of a cmail server system to list, add, modify and delete
user accounts and their login credentials.

.SH OPTIONS
.TP 
.BI "--dbpath " master-database-file ", -d " master-database-file
Select an alternate master database file instead of the default
.IR /etc/cmail/databases/master.db3 .
This overrides the database path provided by the
.B CMAIL_MASTER_DB
environment variable.
.TP
.BI "--verbose " N ", -v " N
Increase verbosity on a scale of
.B 0
to 
.B 4
.TP
.B "--help, -h"
Print helpful hints about operating the program
.SH FILES
.I /etc/cmail/databases/master.db3
.RS
Default path of the master database when installed system-wide. May be overridden either by specifying the
.B CMAIL_MASTER_DB
environment variable or by supplying the
.BR "--dbpath " parameter.
.SH MODES

.TP
.BI "list " [expression]
List all users currently known to the system. Optionally, filter the list for entries matching
.IR expression "."

.TP
.BI "add " "username [password]"
Create a new user and update the login credentials if supplied. User names must be unique.
.RS
.B Example:
$ cmail-admin-user add cbdev donotusethisasyourpassword
.RE

.TP
.BI "revoke " user
Reset the password for the specified user. This causes all login attempts for the user to be
rejected. Note that the system will still accept mail on the users behalf, as
.BR revoke " only invalidates the login credentials, not the account."
.RS
.B Example:
$ cmail-admin-user revoke cbdev
.RE

.TP
.BI "password " "user [password]"
Update the password for the specified user to a new one. If none is supplied, the password is
queried interactively.
.RS
.B Example:
$ cmail-admin-user passwd cbdev anotherbadpassword
.RE

.TP
.BI "alias " "user [alias]"
.RI "Mark " user " as an alias for " alias ". This functionality enables " alias " to have multiple virtual"
.RI "passwords, thus allowing safe sharing of a mailbox. Take note that " user " will only be able to access"
.RI "mail stored to " alias ". This operation can be undone by not specifying the " alias " parameter."
.RS
.B Example:
$ cmail-admin-user alias cbdev cdbev
.RE

.TP
.BI "userdb " "user [path]"
.RI "Have all mail stored for " user " be physically located in an SQLite database at " path ". The user database"
.RI "must conform to the schema outlined in the cmail developer documentation, such as databases created with"
.BR cmail-admin-database (8).
.IR path " should always be an absolute path, as the working directory of processes accessing the user databases"
is subject to change. User databases and the directory containing them must be read- and writable by the user
running the cmail daemons.
.RI "This change can be reversed by not specifying " path ". Note that mail stored in a user database"
that is no longer known to the system can not be retrieved via the cmail suite.
.RS
.B Example:
$ cmail-admin-user userdb cbdev /home/cbdev/mail.db3
.RE

.TP
.BI "delete " user
Delete a specific user
.RS
.B Example:
$ cmail-admin-user delete cbdev
.RE

.SH BUGS
For a discussion of pitfalls and possible sources of misconfigurations refer to the CAVEATS.txt file in the source distribution.
.SH AUTHORS
 Jan "mpease" Duepmeier <me at mpease dot de>  
 Fabian "cbdev" Stumpf <fjs at fabianstumpf dot de>  
.SH "SEE ALSO"
Online documentation and homepage at http://cmail.rocks/
.PP
.BR cmail-admin-address (8),
.BR cmail-admin-smtpd (8),
.BR cmail-admin-popd (8),
.BR cmail-smtpd (8),
.BR cmail-popd (8),
.BR cmail-dispatchd (8),
.BR cmail-master-database (5)



debian/changelog
cmail-mta (1.0) unstable; urgency=low

  * Initial Release.

 -- root <root@unknown>  Sat, 18 Jul 2015 14:24:27 +0000


cmail-admin/user/cmail-admin-user.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sqlite3.h>
#include <time.h>

#define CMAIL_ADMIN_AUTH
#include "../cmail-admin.h"

#include "getpass.c"
#include "user.c"

#define MAX_PW_LENGTH 256
#define PROGRAM_NAME "cmail-admin-user"

int generate_salt(char* out, unsigned chars) {
	const char* salt_alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
	uint8_t randomness[chars];
	int i;

	if(common_rand(randomness, chars) < 0){
		return -1;
	}

	for(i=0;i<chars;i++){
		out[i] = salt_alphabet[randomness[i]%strlen(salt_alphabet)];
	}

	return 0;
}

int set_password(LOGGER log, sqlite3* db, const char* user, char* password) {
	unsigned salt_len = 10;

	char salt[salt_len + 1];
	char password_hash[BASE16_ENCODE_LENGTH(SHA256_DIGEST_SIZE) + 1];
	char auth_data[sizeof(salt) + sizeof(password_hash) + 1];

	memset(salt, 0, sizeof(salt));
	memset(password_hash, 0, sizeof(password_hash));
	memset(auth_data, 0, sizeof(auth_data));

	if(password){
		if(generate_salt(salt, salt_len) < 0){
			logprintf(log, LOG_ERROR, "Failed to generate a random salt\n");
			return 21;
		}

		logprintf(log, LOG_DEBUG, "Generated salt %s\n", salt);

		if(auth_hash(password_hash, sizeof(password_hash), salt, strlen(salt), password, strlen(password)) < 0) {
			logprintf(log, LOG_ERROR, "Error hashing password\n");
			return 21;
		}

		snprintf(auth_data, sizeof(auth_data), "%s:%s", salt, password_hash);
		return sqlite_set_password(log, db, user, auth_data);
	}
	else{
		return sqlite_set_password(log, db, user, NULL);
	}

}

int usage(char* fn){
	printf("cmail-admin-users - Add, modify or delete cmail users\n");
	printf("This program is part of the cmail administration toolkit\n");
	printf("Usage: %s <options> <commands> <arguments>\n\n", fn);

	printf("Basic options:\n");
	printf("\t--verbosity, -v\t\t\tSet verbosity level (0 - 4)\n");
	printf("\t--dbpath, -d <dbpath>\t\tSet master database path (Default: %s)\n", DEFAULT_DBPATH);
	printf("\t--help, -h\t\t\tDisplay this help message\n\n");

	printf("Commands:\n");
	printf("\tlist [<filter>]\t\t\tList all active users, optionally filter by <filter>\n");
	printf("\tadd <username> [<password>]\tAdd an user\n");
	printf("\tpassword <user> [<pw>]\t\tUpdate a users password\n");
	printf("\talias <user> [<alias>]\t\tModify user to be an alias\n");
	printf("\tuserdb <user> [<path>]\t\tSet user database path\n");
	printf("\trevoke <user>\t\t\tRevoke user login credentials (Sets NULL password)\n");
	printf("\tdelete <user>\t\t\tDelete an account\n");
	return 1;
}

#include "../lib/common.c"

int mode_passwd(LOGGER log, sqlite3* db, int argc, char** argv){
	char password[MAX_PW_LENGTH];
	char* pw = password;

	memset(password, 0, sizeof(password));

	if(argc < 2){
		logprintf(log, LOG_ERROR, "Missing user name argument\n");
		return -1;
	}

	else if(argc == 2){
		fprintf(stderr, "Enter new password (leave blank to disable login): ");
		if(ask_password(password, MAX_PW_LENGTH) < 0){
			logprintf(log, LOG_ERROR, "Maximum password length is %d\n", MAX_PW_LENGTH);
			return 11;
		}
	}

	else{
		pw = argv[2];
	}

	return set_password(log, db, argv[1], pw[0] ? pw:NULL);
}

int mode_add(LOGGER log, sqlite3* db, int argc, char** argv){
	if(argc < 2){
		logprintf(log, LOG_ERROR, "Missing user name argument\n\n");
		return -1;
	}
	else{
		if(sqlite_add_user(log, db, argv[1], NULL)){
			logprintf(log, LOG_ERROR, "Failed to create user\n");
			return -1;
		}
		return mode_passwd(log, db, argc, argv);
	}
}

int mode_userdb(LOGGER log, sqlite3* db, int argc, char** argv){
	if (argc < 2){
		logprintf(log, LOG_ERROR, "Missing arguments\n");
	}

	return sqlite_set_userdb(log, db, argv[1], (argc == 3 && strlen(argv[2]) > 1)?argv[2]:NULL);
}

int mode_alias(LOGGER log, sqlite3* db, int argc, char** argv){
	if (argc < 2){
		logprintf(log, LOG_ERROR, "Missing arguments\n");
	}

	return sqlite_set_alias(log, db, argv[1], (argc == 3 && strcmp(argv[1], argv[2]))?argv[2]:NULL);
}

int mode_revoke(LOGGER log, sqlite3* db, int argc, char** argv){
	if(argc != 2){
		logprintf(log, LOG_ERROR, "No user name supplied\n");
		return -1;
	}

	return set_password(log, db, argv[1], NULL);
}

int mode_delete(LOGGER log, sqlite3* db, int argc, char** argv){
	if(argc != 2){
		logprintf(log, LOG_ERROR, "No user name supplied\n");
		return -1;
	}
	return sqlite_delete_user(log, db, argv[1]);
}

int mode_list(LOGGER log, sqlite3* db, int argc, char** argv){
	char* filter = "%";

	if(argc >= 2){
		filter = argv[1];
	}

	return sqlite_get(log, db, filter);
}


int main(int argc, char* argv[]) {
	struct config config = {
		.verbosity = 0,
		.dbpath = getenv("CMAIL_MASTER_DB")
	};

	if(!config.dbpath){
		config.dbpath = DEFAULT_DBPATH;
	}

	// argument parsing
	add_args();

	char* cmds[argc];
	int cmdsc = eargs_parse(argc, argv, cmds, &config);

	LOGGER log = {
		.stream = stderr,
		.verbosity = config.verbosity
	};

	if (cmdsc < 0) {
		return 1;
	}

	int status = -1;

	sqlite3* db = NULL;

	if(!cmdsc){
		logprintf(log, LOG_ERROR, "No command specified\n\n");
		exit(usage(argv[0]));
	}

	logprintf(log, LOG_INFO, "Opening database at %s\n", config.dbpath);
	db = database_open(log, config.dbpath, SQLITE_OPEN_READWRITE);
	if(!db){
		//logprintf(log, LOG_ERROR, "Failed to open database at %s, please check the path\n\n", dbpath);
		exit(usage(argv[0]));
	}

	//check database version
	if (database_schema_version(log, db) != CMAIL_CURRENT_SCHEMA_VERSION) {
		logprintf(log, LOG_ERROR, "The specified database (%s) is at an unsupported schema version.");
		sqlite3_close(db);
		return 11;
	}

	//select command
	if(!strcmp(cmds[0], "add")){
		status = mode_add(log, db, cmdsc, cmds);
	}
	else if(!strcmp(cmds[0], "delete")){
		status = mode_delete(log, db, cmdsc, cmds);
	}
	else if(!strcmp(cmds[0], "revoke")){
		status = mode_revoke(log, db, cmdsc, cmds);
	}
	else if(!strcmp(cmds[0], "list")){
		status = mode_list(log, db, cmdsc, cmds);
	}
	else if(!strcmp(cmds[0], "userdb")){
		status = mode_userdb(log, db, cmdsc, cmds);
	}
	else if(!strcmp(cmds[0], "alias")){
		status = mode_alias(log, db, cmdsc, cmds);
	}
	else if(!strcmp(cmds[0], "password") || !strcmp(cmds[0], "passwd")){
		status = mode_passwd(log, db, cmdsc, cmds);
	}
	else{
		logprintf(log, LOG_WARNING, "Unknown command %s\n\n", cmds[0]);
		usage(argv[0]);
	}

	sqlite3_close(db);
	return status;
}
